File: src/main/java/org/extrasapache/myfaces/codi/examples/ebean/view/person/PersonDetailView.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.extrasapache.myfaces.codi.examples.ebean.view.person;

import org.extrasapache.myfaces.codi.examples.ebean.business.bo.PersonFacade;
import org.extrasapache.myfaces.codi.examples.ebean.orm.Address;
import org.extrasapache.myfaces.codi.examples.ebean.orm.Person;
import org.apache.myfaces.extensions.cdi.core.api.scope.conversation.ViewAccessScoped;

import javax.inject.Inject;
import javax.inject.Named;
import java.io.Serializable;

/**
 * @author werpu2
 */
@Named
@ViewAccessScoped
public class PersonDetailView implements Serializable {
    public static final String NAV_DETAIL = "personDetail";
    public static final String NAV_LIST = "personList";
    private static final String MODE_CREATE = "create";
    private static final String MODE_EDIT = "edit";
    private static final String MODE_DELETE = "delete";

    @Inject
    PersonListSearchModel searchData;

    @Inject
    PersonFacade personFacade;

    Person person;
    Address address;


    String viewMode = MODE_CREATE;

    public String goCreate() {
        viewMode = MODE_CREATE;
        person = personFacade.create();

        return NAV_DETAIL;
    }

    public String goDeta() {
        viewMode = MODE_EDIT;

        return NAV_DETAIL;
    }

    public String goDelete() {
        viewMode = MODE_DELETE;


        return NAV_DETAIL;
    }

    public String doDelete() {
        personFacade.delete(person);

        return NAV_LIST;
    }

    public String doSave() {

        personFacade.save(person);



        return NAV_LIST;
    }

    public String doCancel() {

        personFacade.cancel(person);

        return NAV_LIST;
    }



    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }

    public String getViewMode() {
        return viewMode;
    }

    public void setViewMode(String viewMode) {
        this.viewMode = viewMode;
    }

    public PersonListSearchModel getSearchData() {
        return searchData;
    }

    public void setSearchData(PersonListSearchModel searchData) {
        this.searchData = searchData;
    }

    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }

    public String addAddress() {
        address = personFacade.createAdr();
        person.getAddresses().add(address);

        return NAV_DETAIL;
    }

    public String removeAddress() {
        person.getAddresses().remove(address);

        return NAV_DETAIL;
    }

    //now to the tricky part, in clustered situations the detail view or the session
    //Generally can be serialised and deserialised in this case we lose the entity manager
    //connection
    //since per spec the em hosting ejb never is passivated thanks to the em injected
    //We simply restore the person bean from there, the altered data still should be present
    //todo deserialisation, we reload the person object anew
    /*private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        if(personId != -1L) {
            person = personFacade.loadById(personId);
        }
    } */

}

File: src/main/scala/org/extrasapache/myfaces/codi/examples/ebean/view/person/PersonDetailView.scala
package org.extrasapache.myfaces.codi.examples.ebean.view.person

import org.apache.myfaces.extensions.cdi.core.api.scope.conversation.ViewAccessScoped
import org.apache.myfaces.extensions.cdi.core.api.config.view.ViewConfig
import reflect.BeanProperty
import java.io.Serializable
import javax.inject.{Inject, Named}
import org.extrasapache.myfaces.codi.examples.ebean.business.bo.person.PersonFacade
import org.extrasapache.myfaces.codi.examples.ebean.orm.person.Address

object PersConst {
  val MODE_CREATE = "create"
  val MODE_EDIT = "edit"
  val MODE_DELETE = "delete"
}


/**
 * PersonListViewModel
 * we split our page bean into two parts
 * a model part containing all referenced
 * model elements (and combined controller model elements)
 *
 * and a controller class part containing the core logic
 * That way we get a cleaner structure  and can share models
 * between various page beans with different logic
 *
 */
trait PersonDetailViewModel  {

  @Inject
  @BeanProperty
  var searchData: PersonListSearchModel = _

  @Inject
  var personFacade: PersonFacade = _

  @BeanProperty
  var person: org.extrasapache.myfaces.codi.examples.ebean.orm.person.Person = _

  @BeanProperty
  var address: Address = _

  @BeanProperty
  var viewMode: String = PersConst.MODE_CREATE
}

/**
 * The controller logic which itself is also a page bean,
 * we use traits as decorators which decorate our model in
 */
@Named
@ViewAccessScoped
class PersonDetailView extends PersonDetailViewModel with Serializable {

  def goCreate: java.lang.Class[_ <: ViewConfig] = {
    viewMode = PersConst.MODE_CREATE
    person = personFacade.create

    classOf[Person.PersonDetail]
  }

  def goDeta: java.lang.Class[_ <: ViewConfig] = {
    viewMode = PersConst.MODE_EDIT
    classOf[Person.PersonDetail]
  }

  def doDelete: java.lang.Class[_ <: ViewConfig] = {
    personFacade.delete(person)
    classOf[Person.PersonList]
  }

  def doSave: java.lang.Class[_ <: ViewConfig] = {
    personFacade.save(person)
    classOf[Person.PersonList]
  }

  def doCancel: java.lang.Class[_ <: ViewConfig] = {
    personFacade.cancel(person)
    classOf[Person.PersonList]
  }

  def addAddress: java.lang.Class[_ <: ViewConfig] = {
    address = personFacade.createAdr
    person.getAddresses.add(address)

    classOf[Person.PersonDetail]
  }

  def removeAddress: java.lang.Class[_ <: ViewConfig] = {
    person.getAddresses.remove(address)

    classOf[Person.PersonDetail]
  }

}